ion-schema/src/isl/isl_type_reference.rs

```diff
@@ -208,20 +207,21 @@ impl IslTypeRefImpl {
                 if value_struct.get("id").is_none() {
                     let type_def= IslTypeImpl::from_owned_element(isl_version, value, inline_imported_types)?;
                     // if type reference contains `occurs` field and has modifier `$null_or` then return an error
-                    if nullability == NullOr && type_def.constraints()
-                        .iter()
-                        .any(|c| matches!(c, IslConstraintImpl::Occurs(_))) {
+                    if nullability == NullOr && value_struct.get("occurs").is_some() {
                         return invalid_schema_error(
                             "`$null_or` annotation is not supported for a type reference with an explicit `occurs` field",
                         )
                     }
 
-                    if !allow_occurs_field && isl_version == IslVersion::V2_0 && type_def.constraints()
-                        .iter()
-                        .any(|c| matches!(c, IslConstraintImpl::Occurs(_))) {
-                        return invalid_schema_error(
+                    if let Some(occurs_constraint) = value_struct.get("occurs") {
```

Comment:
Little nitpick about `occurs` not being a constraint—maybe use `occurs_field` instead of `occurs_constraint`.